src/hypofuzz/database.py

```diff
@@ -41,9 +51,68 @@ class WorkerIdentity:
     git_hash: Optional[str]
 
     @staticmethod
-    def from_dict(data: dict) -> "WorkerIdentity":
+    def from_json(data: dict) -> "WorkerIdentity":
         return WorkerIdentity(**data)
 
+    def __repr__(self) -> str:
+        return f"WorkerIdentity(uuid={self.uuid!r})"
+
+    __str__ = __repr__
+
+
+class ObservationStatus(Enum):
+    PASSED = "passed"
+    FAILED = "failed"
+    GAVE_UP = "gave_up"
+
+
+@dataclass(frozen=True)
+class Observation:
+    # attributes from
+    # https://hypothesis.readthedocs.io/en/latest/reference/integrations.html#test-case
+
+    # we're only storing test_case reports for now. We may store information
+    # messages in the future as well.
+    type: Literal["test_case"]
+    status: ObservationStatus
+    status_reason: str
+    representation: str
+    arguments: dict[str, Any]
+    how_generated: str
+    features: dict[str, Any]
+    timing: dict[str, Any]
+    metadata: dict[str, Any]
+    property: str
+    run_start: int
+
+    @staticmethod
+    def from_json(data: dict) -> Optional["Observation"]:
+        data = dict(data)
+        # we disable observability coverage, but hypothesis still gives us a key
+        # for it.
+        if "coverage" in data:
+            data.pop("coverage")
+        try:
+            data["status"] = ObservationStatus(data["status"])
+            return Observation(**data)
+        except Exception:
+            return None
+
+
+@dataclass(frozen=True)
+class FailureRepresentation:
+    traceback: str
+    call_repr: str
+    reproduction_decorator: str
+
+    @staticmethod
+    def from_json(data: dict) -> Optional["FailureRepresentation"]:
+        data = dict(data)
+        try:
+            return FailureRepresentation(**data)
+        except Exception:
+            return None
```

Comment:
as discussed: we can have Hypothesis store the `@reproduce_failure(...)` string in the observations, and store the traceback for _all_ failures, at which point we won't need this separate class.